### ISSUE TITLE
Exit code 404

### ISSUE THREAD

**Issue by demdante** [chocolatey/package-verifier#88]
Hello.  I recently submitted a chocolatey package that is failing the automatic verification process.  The test environment is returning error code 404: ERROR: The remote file either doesn't exist, is unauthorized, or is forbidden for url 'https://www.secretptr.com/rc_images/secret_windows_1_0.zip'. Exception calling "GetResponse" with "0" argument(s): "The request was aborted: Could not create SSL/TLS secure channel.  However, when I tested the package in my test vm, the package successfully installed.  It would appear that the test environment needs to be upgraded, as it is failing to create ssl/tls secure channels.  

![image](https://user-images.githubusercontent.com/64720591/88769590-207da600-d16c-11ea-952d-a2950ed03a76.png)

![image](https://user-images.githubusercontent.com/64720591/88770289-1f00ad80-d16d-11ea-86b4-9724863f99ee.png)

**Comment by ferventcoder** [comment]
@demdante This is likely the URL it's reaching out to requires updated ciphers, which are brought in the way of a KB that you'll need to add as a dependency (and then request a package verifier exception)

* https://www.ssllabs.com/ssltest/analyze.html?d=www.secretptr.com&latest

![image](https://user-images.githubusercontent.com/63502/89071702-d8fa4400-d33c-11ea-97ce-27c05eae7dd0.png)


@pauby may have already addressed this at https://gitter.im/chocolatey/choco?at=5f217414d7efe5438f2b9a90:

 > While Chocolatey uses TLS 1.2, Windows Server 2012R2 doesn't have the ciphers installed to allow it to happen at operating system level. So that KB2919355 patch is required to add those ciphers. If you do add that as a dependency, request an exemption from testing as the patch needs a reboot to take effect.
Apologies for the slight misinformation. All my fault.

Basically it means the package needs to take a dependency on the `KB2919355` package.

**Comment by ferventcoder** [comment]
I'll go ahead and close this out.

While one might initially believe something like this means the test environment should be upgraded, if you spin that on its head and think about if I was to start from a base install and install this package, would it fail or would it succeed? 

In that aspect, we want to capture all required needs for a package as far as dependencies go, even OS level. HTH

**Comment by demdante** [comment]
@ferventcoder Ok.  So, let me make sure I understand what you are telling me is the solution to this issue.  You are telling me that I need to add an **operating system specific windows update** as a **universal dependency** in the nuspec file for the chocolatey package in question.  I understand where you are coming from and even if it were clear in the first place how to designate a windows update msu as a dependency in the nuspec file, what you are saying is the solution would break the chocolatey package for all but server 2012 R2/8.1 devices, which I think is much worse than just updating the test environment.  Or, were you suggesting manually performing a dependency validation/install within the chocolateyinstall script, which, while the scripting is not difficult, means I have to research for the KB with the same updates for other operating systems as well, and then it spirals into a needlessly complicated mess.  I think this is just one of those things where, if users want the product to work correctly, then, like for most things, they need to make sure that they have the appropriate updates installed.

**Comment by pauby** [comment]
@demdante I appreciate you've responded to @ferventcoder but I wanted to jump in and say [KB2919355 is already a package](https://chocolatey.org/packages/KB2919355) so nothing needs to be created. You simply need to add it as a dependency.

**Comment by ferventcoder** [comment]
> what you are saying is the solution would break the chocolatey package for all but server 2012 R2/8.1 devices

Thankfully I was not saying this at all 😄 . The package that handles the dependency will make a decision on what it needs to do (what @pauby mentioned), and in quite a few cases will not do anything.

> means I have to research for the KB with the same updates for other operating systems as well, and then it spirals into a needlessly complicated mess

Yes, if you were the first person to come across this aspect, it would be expected to create the package that handles that, but fortunately many of these lower-level dependencies already have packages surrounding them. So all you need is to add a line or two to the nuspec to handle the dependency on the package.

Most of the logic is handled at https://chocolatey.org/packages/KB2919355#files (chocolateyInstall.ps1): 

```powershell
# [..snip..]
# more code above 

if ($version -eq $null -or $version -lt [Version]'6.3' -or $version -ge [Version]'6.4') {
	Write-Host "Skipping installation because this hotfix only applies to Windows 8.1 and Windows Server 2012 R2."
	return
}

if (Get-HotFix -id $kb -ea SilentlyContinue)
{
	Write-Host "Skipping installation because hotfix $kb is already installed."
	return
}

# other code below, installing the MSU
# [..snip..]

```

(these are lower level packages, so the idea is to do nothing with a message

**Comment by demdante** [comment]
@ferventcoder and @pauby  So I guess I am getting confused as to why you keep referencing KB2919355 as the universal dependency required, when it is only applicable to windows 8.1/server 2012 R2.  When I said "means I have to research for the KB with the same updates for other operating systems as well, and then it spirals into a needlessly complicated mess," based on the code snippet you provided, you seem to think I was referring to the different os variants to which that Update is applicable, which is not what I was trying to say.  What I was referring to were older operating systems, like windows 7 and server 2008 R2.  Chocolatey can run on them, and it sounds like you want supported operating systems to be able to run packages right out of the box, right?

> While one might initially believe something like this means the test environment should be upgraded, if you spin that on its head and think about if I was to start from a base install and install this package, would it fail or would it succeed?
> 
> In that aspect, we want to capture all required needs for a package as far as dependencies go, even OS level. HTH

That is what I am saying: assuming that operating systems newer than server 2012 R2 contain the tls ciphers in the base OS and require no update, what you are suggesting still leaves out Windows 2008 R2 and Windows 7, or do we not care about those operating systems?  If not, then do we only care about the package-verifier operating system and newer?  If so, then why?  But we already know the answer to that: because the amount of work involved in preparing a windows 7 base install to even be able to download files from the internet using powershell is far beyond the capabilities of a single chocolatey package.  In fact, if memory serves, base installs of windows 7 can't even successfully install chocolatey using the command line installation.  And just to be clear, I am not saying that dependency checking is bad, when the dependencies are external pieces of software that are not os dependent and not built into the operating system.  But when you enforce windows updates as dependencies (considering that users who are properly securing their machines **should** already have up-to-date systems) at what point does the operating system become so old and the amount of patching so great that the "base operating system compatibility" is not feasible and doesn't matter (kind of like chocolatey itself and windows 7)?  Is it at the point between server 2008 R2 and server 2012?  And if new technologies in the future are introduced that make a base install of server 2012 R2 as difficult to install dependencies for as server 2008 R2 is today?  The point that I am trying to make here is that your goal of requiring that packages work right even on base install operating systems is unrealistic.  Just my two cents on the matter, as I have no intention of committing any more time to the package just to provide compatibility for out-of-date server 2012 R2 machines.

**Comment by pauby** [comment]
@demdante I'm going to put my hand up here and say I'm also getting a little confused. I think we're coming at it from two different positions and assuming the other gets where we are coming from. I'll try to answer your questions though.

> @ferventcoder and @pauby So I guess I am getting confused as to why you keep referencing KB2919355 as the universal dependency required, when it is only applicable to windows 8.1/server 2012 R2. 

It is only required for those operating systems.

> When I said "means I have to research for the KB with the same updates for other operating systems as well, and then it spirals into a needlessly complicated mess," based on the code snippet you provided, you seem to think I was referring to the different os variants to which that Update is applicable, which is not what I was trying to say. What I was referring to were older operating systems, like windows 7 and server 2008 R2. Chocolatey can run on them, and it sounds like you want supported operating systems to be able to run packages right out of the box, right?

Chocolatey no longer supports Windows 7 and only supports Windows 2008R2 on Azure. See [this blog post](https://chocolatey.org/blog/remove-support-for-old-tls-versions), particularly the sections on 'Why are we doing this now?' and 'Provisioning Older Machines'. So, in short you don't have to worry about those.

The rest of your comments seem to centre around Windows 7 and 2008R2 support which I hope has been answered by the above? Let me know if that is not the case.
